tests/src/tests/command_test_utils.go

```diff
@@ -0,0 +1,258 @@
+package tests
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/google/go-cmp/cmp"
+	"sigs.k8s.io/yaml"
+
+	schema "github.com/devfile/api/pkg/apis/workspaces/v1alpha2"
+)
+
+// Return a specifed number of env attributes in a schema structure
+func AddEnv(numEnv int) []schema.EnvVar {
+	commandEnvs := make([]schema.EnvVar, numEnv)
+	for i := 0; i < numEnv; i++ {
+		commandEnvs[i].Name = "Name_" + GetRandomString(5, false)
+		commandEnvs[i].Value = "Value_" + GetRandomString(5, false)
+		LogInfoMessage(fmt.Sprintf("Add Env: %s", commandEnvs[i]))
+	}
+	return commandEnvs
+}
+
+// Return a specifed number of attributes in a schema structure
+func AddAttributes(numAtrributes int) map[string]string {
+	attributes := make(map[string]string)
+	for i := 0; i < numAtrributes; i++ {
+		AttributeName := "Name_" + GetRandomString(6, false)
+		attributes[AttributeName] = "Value_" + GetRandomString(6, false)
+		LogInfoMessage(fmt.Sprintf("Add attribute : %s = %s", AttributeName, attributes[AttributeName]))
+	}
+	return attributes
+}
+
+// Create and return a group in a schema structure
+func addGroup() *schema.CommandGroup {
+
+	commandGroup := schema.CommandGroup{}
+	commandGroup.Kind = GetRandomGroupKind()
+	LogInfoMessage(fmt.Sprintf("group Kind: %s", commandGroup.Kind))
+	commandGroup.IsDefault = GetBinaryDecision()
+	LogInfoMessage(fmt.Sprintf("group isDefault: %t", commandGroup.IsDefault))
+	return &commandGroup
+}
+
+// Add a command of the specified type to the schema
+func (devfile *TestDevfile) addCommand(commandType schema.CommandType) string {
+	command := generateCommand(commandType)
+	devfile.SchemaDevFile.Commands = append(devfile.SchemaDevFile.Commands, command)
+	return command.Id
+}
+
+// Create  a command of a specified type in a schema structure
+func generateCommand(commandType schema.CommandType) schema.Command {
+	command := schema.Command{}
+	command.Id = GetRandomUniqueString(8, true)
+	LogInfoMessage(fmt.Sprintf("command Id: %s", command.Id))
+
+	if commandType == schema.ExecCommandType {
+		command.Exec = createExecCommand()
+	} else if commandType == schema.CompositeCommandType {
+		command.Composite = createCompositeCommand()
+	}
+	return command
+}
+
+// Update the values of a specified command
+func (devfile *TestDevfile) UpdateCommand(parserCommand *schema.Command) error {
+
+	var errorString []string
+	testCommand, found := getSchemaCommand(devfile.SchemaDevFile.Commands, parserCommand.Id)
```

Comment:
why are we doing this? You already have the command that you passed in to this function - `parserCommand`. The `getSchemaCommand()` basically loops thru all the commands and matches the id and returns the same command back..

Reply:
So here I am testing the paser api used to modify existing devfile conent. So first I create the schema structure as reference devfile content. Write that to disk, have the parser read it and then start updating. For updating I update the reference structure, use parser api to make the same update, get the full devfile back and make sure the updates are reflected.  Probably need to get some invalid updates going here to that is for the future.

Reply:
I get what you mean, but you are not doing anything with the `testCommand` in this func:

```
var err error
	testCommand, found := getSchemaCommand(devfile.SchemaDevFile.Commands, parserCommand.Id)
	if found {
		LogInfoMessage(fmt.Sprintf("Updating command id: %s", parserCommand.Id))
		if testCommand.Exec != nil {
			setExecCommandValues(parserCommand.Exec)
		} else if testCommand.Composite != nil {
			setCompositeCommandValues(parserCommand.Composite)
		}
		devfile.replaceSchemaCommand(*parserCommand)
	} else {
		err = errors.New(LogErrorMessage(fmt.Sprintf("Command not found in test : %s", parserCommand.Id)))
	}
	return err
```

Reply:
testCommand is used in the if and else statements. The logic is basically

1. Use the id from the command returned by the parser to find the equivalent command in the test.
2: if it is found:
    a. check the type of the command recorded in the test and set appropriate update values in the command returned by the parser
   b. replace the command stored in the test with the new version to be given to the parser.
3. If it is not found we have a problem.

on return the parser api is use to write the updated parser command.

I could do this any number of ways but the outcome needs to be the same.